File: plugin/com.hello.uxml.tools.ui/src/com/hello/uxml/tools/framework/Group.java
package com.hello.uxml.tools.framework;

/**
 * Implements a stack container for elements.
 *
 * @author ferhat
 */
public class Group extends UIElementContainer {
  @Override
  protected void onMeasure(double availableWidth, double availableHeight) {
    int childCount = getChildCount();
    if (childCount == 0) {
      setMeasuredDimension(0, 0);
    } else {
      UIElement child = getChild(0);
      child.measure(availableWidth, availableHeight);
      double maxWidth = child.getMeasuredWidth();
      double maxHeight = child.getMeasuredHeight();
      for (int c = 1; c < childCount; ++c) {
        child = getChild(c);
        child.measure(availableWidth, availableHeight);
        if (child.getMeasuredWidth() > maxWidth) {
          maxWidth = child.getMeasuredWidth();
        }
        if (child.getMeasuredHeight() > maxHeight) {
          maxHeight = child.getMeasuredHeight();
        }
      }
      setMeasuredDimension(maxWidth, maxHeight);
    }
  }

  @Override
  protected void onLayout(Rectangle layoutRect) {
    int childCount = getChildCount();
    Rectangle childRect = new Rectangle(0, 0, layoutRect.width, layoutRect.height);
    for (int c = 0; c < childCount; ++c) {
      getChild(c).layout(childRect);
    }
  }
}

File: framework/lib/group.dart
part of uxml;

/**
 * Groups UIElement's as a stack.
 *
 * @author ferhat@ (Ferhat Buyukkokten)
 */
class Group extends UIElementContainer {
  static ElementDef groupElementDef;

  Group() : super() {
  }

  /** Overrides UIElement.onMeasure to return size of element. */
  bool onMeasure(num availableWidth, num availableHeight) {
    int numChildren = childCount;
    num maxWidth = 0.0;
    num maxHeight = 0.0;
    for (int i = 0; i < numChildren; i++) {
      UIElement child = childElements[i];
      child.measure(availableWidth, availableHeight);
      if (child.measuredWidth > maxWidth) {
        maxWidth = child.measuredWidth;
      }
      if (child.measuredHeight > maxHeight) {
        maxHeight = child.measuredHeight;
      }
    }
    setMeasuredDimension(maxWidth, maxHeight);
    return false;
  }

  /** Overrides UIElement.onLayout. */
  void onLayout(num targetX, num targetY, num targetWidth, num targetHeight) {
    int numChildren = childCount;
    for (int i = 0; i < numChildren; i++) {
      childAt(i).layout(0.0, 0.0, targetWidth, targetHeight);
    }
  }

  /** @see UxmlElement.getDefinition. */
  ElementDef getDefinition() => groupElementDef;

  /** Registers component. */
  static void registerGroup() {
    groupElementDef = ElementRegistry.register("Group",
        UIElementContainer.uielementcontainerElementDef, null, null);
  }
}
